### ISSUE TITLE
Space complexity & Time Complexity

### ISSUE THREAD

**Issue by melkir** [renshuki/fibonacci-ruby#1]
Fibonacci time complexity:
[Recursive](https://stackoverflow.com/a/360773/3495440): O(2^n)
[Iterative](https://stackoverflow.com/a/13440108/3495440): ~~O(2^n)~~ 
[Memoization](https://stackoverflow.com/a/13440108/3495440): O(n)

Fibonacci space complexity : 
[Recursive](https://stackoverflow.com/a/360773/3495440): O(n)
[Iterative](https://stackoverflow.com/a/360773/3495440): O(1)
[Memoization](https://stackoverflow.com/a/13440108/3495440): O(n)

[Explaination why iterative is better than recursive](https://stackoverflow.com/a/8534328/3495440)

Correct me if I'm wrong :)

Also please consider adding the power of matrix implementation so we can have a O(log(n)) time, a very good resource : http://www.geeksforgeeks.org/program-for-nth-fibonacci-number/

**Comment by renshuki** [comment]
Thank you for the issue ;)
I didn't read all in details but it looks good to me :+1: 

> Also please consider adding the power of matrix implementation so we can have a O(log(n)) time, a very good resource : http://www.geeksforgeeks.org/program-for-nth-fibonacci-number/

Please feel free to do a PR for that, it would be interesting to see matrix performance comparison  :pray:

**Comment by renshuki** [comment]
I just had to remember some BigO stuff this afternoon and I took a look to your issue again.  
IMO the time complexity for the iteration seems to be O(n) (linear) not O(2^n).

**Comment by melkir** [comment]
You're right except if you use some BigNum representation because 64 bit integer may be not enough (starting from k=94) which add to the complexity of the algorithm.
https://www.quora.com/What-is-the-time-complexity-for-an-iterative-solution-to-Fibonacci-series

**Comment by renshuki** [comment]
Yes, but in this case I assume that time for iterative is growing exponentially so it's **O(n^2)** not **O(2^n)** right?  
Or there is something I don't get 😭

**Comment by melkir** [comment]
Yup, O(n^2) good catch ;)
To summarize the iterative solution is O(n) for a small number and O(n^2) for a large number

**Comment by renshuki** [comment]
Hehe np, it was great to have the chance to talk about that with you.
Thanks :)

**Comment by yoadwo** [comment]
Yeah, for recursive space complexity, I actually prefer to treat the space complexity is O(h) where h is depth of tree

**Comment by renshuki** [comment]
@yoadwo It's a good idea for the differentiation 😄 
Btw I found this super handy _Big-O Cheat Sheet_ recently:
http://bigocheatsheet.com/
